File: 48.旋转图像.py
#
# @lc app=leetcode.cn id=48 lang=python3
#
# [48] 旋转图像
#

# @lc code=start
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix[0])
        for i in range (0, n-1):
            for j in range(0, n-1):
                if i + j < n-1:
                    new_i = n - j - 1
                    new_j = n - i - 1
                    matrix[i][j], matrix[new_i][new_j] = matrix[new_i][new_j], matrix[i][j]
        
        for i in range(0, n // 2):
            matrix[i], matrix[n-i-1] = matrix[n-i-1], matrix[i]

# @lc code=end

File: 48.旋转图像.java
/*
 * @lc app=leetcode.cn id=48 lang=java
 *
 * [48] 旋转图像
 */

// @lc code=start
class Solution {
    public void rotate(int[][] matrix) {
        // 单次循环中旋转4个矩形
        int n = matrix.length;
        for (int i = 0; i <(n+1)/2; i++) {
            for (int j = 0; j < n/2; j++) {
                int temp = matrix[n-1-j][i];
                matrix[n-1-j][i] = matrix[n-1-i][n-j-1];
                matrix[n-1-i][n-j-1] = matrix[j][n-1-i];
                matrix[j][n-1-i] = matrix[i][j];
                matrix[i][j] = temp;
            }
        }

        // 旋转四个矩形
        // int n = matrix.length;
        // for (int i = 0; i < n/2 + n%2; i++) {
        //     for (int j = 0; j < n/2; j++) {
        //         int[] tmp = new int[4];
        //         int row = i; 
        //         int col = j;
        //         for (int k = 0; k < 4; k++) {
        //             tmp[k] = matrix[row][col];
        //             int x = row;
        //             row = col;
        //             col = n - 1 - x;
        //         }
        //         for (int k = 0; k < 4; k++) {
        //             matrix[row][col] = tmp[(k+3)%4];
        //             int x = row;
        //             row = col;
        //             col = n - 1 - x;
        //         }
        //     }
            
        // }


        // // 转置加翻转
        // int n = matrix.length;

        // // transpose matrix
        // for (int i = 0; i < n ;i++) {
        //     for (int j = i; j < n; j++) {
        //         int tmp = matrix[j][i];
        //         matrix[j][i] = matrix[i][j];
        //         matrix[i][j] = tmp;
        //     }
        // }
        // // reverse each row
        // for (int i = 0; i < n; i++) {
        //     for (int j = 0; j < n/2; j++) {
        //         int tmp = matrix[i][j];
        //         matrix[i][j] = matrix[i][n-j-1];
        //         matrix[i][n-j-1] = tmp;
        //     }
        // }

    }
}
// @lc code=end

